### PR TITLE
allow_nondeterministic_mutations

### LINKED ISSUE #10185: ALTER UPDATE/ALTER DELETE statements must use only deterministic functions! Function 'dictHas' is non-deterministic
After #7247 certain scenarios of mutations with dictionaries did stop working, and send error messages like "Function 'dictHas' is non-deterministic".

Говорят непонятно как теперь апдейты делать.

> Здравствуйте! Вы мне както подсказывали как в кликхаусе делаете мутации с условиями через словари:
> 
> Denny Crane [not a Yandex bot], [22.01.20 04:27]
> такие запросы нельзя делать, alter для каждой строки xxx проверить where.
> 
> Я готовлю таблицу ReplicatedMergeTree c id-шками которые надо удалить, она обязательно должна создержать одинаковые записи на всех репликах, чтобы мутации на репликах получили один и тот же результат мутации.
> 
> Создаю словарь который смотрит на ReplicatedMergeTree на всех репликах
> 
> Выясняю какие парты содержат записи которые надо удалить
> 
> Выполняю alter table xxx delete where date between парты and dictHas(словарь)
> 
> 
> у меня это прекрасно работало когда кх был сингл инстанс, а сейчас когда появилась реплика, ругается на то, что :
> 
> DB::Exception: ALTER UPDATE/ALTER DELETE statements must use only deterministic functions! Function 'dictHas' is non-deterministic (version 19.17.6.36 (official build))
> 
> создал все как вы советовали таблицы и словари синхронизированы на обоих репликах и имеют одинаковое кол-во записей
> 
> не могли бы вы подсказать в чем может быть проблема?

Я тоже делаю все мутации через словари. К счастью в 18.14.18 такой проблемы нет.

_Originally posted by @den-crane in https://github.com/ClickHouse/ClickHouse/pull/7247#issuecomment-606806745_

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add the ability to relax the restriction on non-deterministic functions usage in mutations with `allow_nondeterministic_mutations` setting.  

Detailed description / Documentation draft:
By default on ReplicatedMergeTree mutations which use non-deterministic functions are forbidden (see #7247), because it may lead to checksum mismatches and redownloading of mutated parts.
But as that is was working before and there are situations when non-deterministic behaves deterministically (for example properly loaded and synced dictionary). we add a setting called 
`allow_nondeterministic_mutations` to allow that behavior.

Usage:

For single query:
```sql
ALTER TABLE table_name DELETE WHERE dictHas(...) SETTINGS allow_nondeterministic_mutations=1;
```

For all queries - add <allow_nondeterministic_mutations>1</allow_nondeterministic_mutations> the the default profile in users.xml

Closes #10185

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@filimonov FYI @blinkov is actually a robot.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -418,7 +418,7 @@ struct Settings : public SettingsCollection<Settings>
     M(SettingBool, merge_tree_uniform_read_distribution, true, "Obsolete setting, does nothing. Will be removed after 2020-05-20", 0) \
     M(SettingUInt64, mark_cache_min_lifetime, 0, "Obsolete setting, does nothing. Will be removed after 2020-05-31", 0) \
     M(SettingBool, partial_merge_join, false, "Obsolete. Use join_algorithm='prefer_partial_merge' instead.", 0) \
-
+    M(SettingBool, allow_nondeterministic_mutations, false, "Allow non-deterministic functions in ALTER UPDATE/ALTER DELETE statements", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:421]: (minor) You have appended it to a group of settings marked obsolete.

**Review by alexey-milovidov — APPROVED** [review]
LGTM

File: src/Core/Settings.h (hunk 1 above)
Comment (filimonov) [reply src/Core/Settings.h]: Ok. Just added at the end, didn't check the comment above :)

File: tests/queries/0_stateless/01017_mutations_with_nondeterministic_functions_zookeeper.sh [hunk 2]
```diff
@@ -35,9 +51,21 @@ ${CLICKHOUSE_CLIENT} --query "ALTER TABLE $T1 DELETE WHERE rand() = 0" 2>&1 > /d
 ${CLICKHOUSE_CLIENT} --query "ALTER TABLE $T1 UPDATE y = y + rand() % 1 WHERE not ignore()" 2>&1 > /dev/null \
 && echo 'OK' || echo 'FAIL'
 
+# hm... it looks like joinGet condidered determenistic
+${CLICKHOUSE_CLIENT} --query "ALTER TABLE $R1 UPDATE y = joinGet('${CLICKHOUSE_DATABASE}.lookup_table', 'y_new', y) WHERE x=1" 2>&1 \
```
Comment (azat) [inline tests/queries/0_stateless/01017_mutations_with_nondeterministic_functions_zookeeper.sh:55]: Hm, but why, maybe it should be marked as non-deterministic instead? Objections?

**Pull request merged**